### PR TITLE
feat(docker): Added support for known ECR tokens (WIP)

### PR DESCRIPTION
**Pull request by jheddings**
## Changes

This PR adds support for known ECR tokens.  Using this approach, CI tools are able to pass a known token directly via Basic auth on ECR repositories.

## Context

Our CI toolchain obtains an ECR token for multiple operations where using Key ID and Secret are not sufficient.

Related Issues:
- renovatebot/renovate#11322
- renovatebot/renovate#11325

Discussion:
- https://github.com/renovatebot/renovate/discussions/20326
- https://github.com/renovatebot/renovate/discussions/18755

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/docker/index.ts [hunk 1]
```diff
@@ -244,6 +244,9 @@ async function getECRAuthToken(
       secretAccessKey: opts.password,
       ...(opts.token && { sessionToken: opts.token }),
     };
+  } else if (opts.token) {
+    addSecretForSanitizing(opts.token);
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:248]:
```suggestion
```
all hostrules are already sanitized

**Review by viceice — COMMENTED** [review]
needs a unit test

**Review by viceice — COMMENTED** [review]
please use trace log level, otherwise we get too many message

**Comment by jheddings** [comment]
> needs a unit test

I'll try to work up a unit test for the new auth methods...  I'm not familiar with Jest, so it may take some time and any help would be appreciated!

File: docs/usage/docker.md [hunk 2]
```diff
@@ -248,7 +248,12 @@ module.exports = {
 
 #### AWS ECR (Amazon Web Services Elastic Container Registry)
 
-Renovate can authenticate with AWS ECR using AWS access key id & secret as the username & password, for example:
+Renovate can authenticate with AWS ECR using AWS credentials, login password or an authorization token.
+
+##### Using AWS account credentials
+
+To authenticate using AWS credentials, specify the access key id & secret as the username & password. If required
+define the session token also, for example:
```
Comment (HonkingGoose) [inline docs/usage/docker.md:256]:
```suggestion
To authenticate using AWS credentials, you must use:

- the access key id as your username
- the secret as your pasword

If needed, you can set a session token, for example:
```

Please check if this suggestion is correct before applying it. The idea here is to put the terms next to their replacements, this should help users to use the right term in the right place.

How does a Renovate user know "If required define the session token also" applies to them? Do we need to explain this in the Renovate docs maybe?

File: docs/usage/docker.md [hunk 3]
```diff
@@ -265,6 +270,46 @@ Renovate can authenticate with AWS ECR using AWS access key id & secret as the u
 }
 ```
 
+##### Using the login password (basic auth)
+
+To authenticate using basic auth, simply provide the password in your config. For example:
```
Comment (HonkingGoose) [inline docs/usage/docker.md:275]:
```suggestion
To authenticate with basic authentication, put the encrypted password in your config.
For example:
```

File: docs/usage/docker.md [hunk 4]
```diff
@@ -265,6 +270,46 @@ Renovate can authenticate with AWS ECR using AWS access key id & secret as the u
...
+      "matchHost": "12345612312.dkr.ecr.us-east-1.amazonaws.com",
+      "encrypted": {
+        "password": "w...A"
+      }
+    }
+  ]
+}
+```
+
+To obtain a login password, using the command `aws ecr get-login-password` or the corresponding API.
```
Comment (HonkingGoose) [inline docs/usage/docker.md:291]:
```suggestion
To get a login password, use the command `aws ecr get-login-password` or the corresponding API.
```

Do we need to link to the API docs for this? Or do we expect AWS users to know this already?

File: docs/usage/docker.md [hunk 5]
```diff
@@ -265,6 +270,46 @@ Renovate can authenticate with AWS ECR using AWS access key id & secret as the u
...
+
+##### Using an authorization token
+
+To authenticate using an authorization token, provide only the token in your config. For example:
```
Comment (HonkingGoose) [inline docs/usage/docker.md:295]:
```suggestion
To authenticate with an authorization token, put the token in your config.
For example:
```

File: docs/usage/docker.md [hunk 6]
```diff
@@ -265,6 +270,46 @@ Renovate can authenticate with AWS ECR using AWS access key id & secret as the u
...
+      "matchHost": "12345612312.dkr.ecr.us-east-1.amazonaws.com",
+      "encrypted": {
+        "token": "QV...09"
+      }
+    }
+  ]
+}
+```
+
+To obtain an authorization token, using the command `aws ecr get-authorization-token` or the corresponding API.
```
Comment (HonkingGoose) [inline docs/usage/docker.md:311]:
```suggestion
To get an authorization token, use the command `aws ecr get-authorization-token` or the corresponding API.
```

Same question here, do we expect AWS users to know about the API, or should we link to the docs?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/docker.md (hunk 4 above)
Comment (jheddings) [reply docs/usage/docker.md:291]: I would generally assume that anyone using Codeartifact would know about the AWS credentials for CLI or API use.  We can certainly link to the docs if it helps with clarity:
https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html

File: docs/usage/docker.md (hunk 2 above)
Comment (jheddings) [reply docs/usage/docker.md:256]: The suggestion looks good, thanks!

The token requirement depends on how the AWS credentials are issued...  It is possible to issue credentials both with and without requiring a session token.  It is typically managed by the AWS account admin.

**Comment by rarkins** [comment]
Closing due to inactivity

**Pull request closed without merging**